04 - Poker.playground/Contents.swift

```diff
@@ -1,32 +1,412 @@
+//: Playground - noun: a place where people can play
+
+import UIKit
 
- - Wir modellieren `Card` als Struct, und `Suit` und `Rank` als Enums. Warum verwenden wir keine Klasse für `Card`? Warum eignet sich ein Enum so hervorragend für `Suit` und `Rank`?
-*/
 /*
- EURE ANTWORT HIER
-*/
+ Enum für Farben
+ */
+enum Suit : Int {
+    case Karo = 1, Herz = 2, Pik = 3, Kreuz = 4
+    var symbol: String {
+        switch(self.rawValue) {
+        case 1: return "\u{2662}"
+        case 2: return "\u{2661}"
+        case 3: return "\u{2664}"
+        case 4: return "\u{2667}"
+        default: return "Something went wrong"
+        }
+    }
+}
+
 
 
+/*
+ Enum für Werte der Karten
+ -> 0 für Vergleichszwecke
+ */
+enum Rank : Int {
+    case Nothing = 0, Zwei = 1, Drei = 2, Vier = 3, Fünf = 4, Sechs = 5, Sieben = 6, Acht = 7, Neun = 8, Zehn = 9, Bube = 10, Dame = 11, König = 12, Ass = 13
+    var symbol: String {
+        switch(self.rawValue) {
+        case 0: return "Nichts"
+        case 1: return "2"
+        case 2: return "3"
+        case 3: return "4"
+        case 4: return "5"
+        case 5: return "6"
+        case 6: return "7"
+        case 7: return "8"
+        case 8: return "9"
+        case 9: return "10"
+        case 10: return "B"
+        case 11: return "D"
+        case 12: return "K"
+        case 13: return "A"
+        default: return "Something went wrong"
+        }
+    }
+}
+
+
+
+/*
+ Struct für Karten mit Vergleichsoption
+ */
+struct Card : Equatable {
+    let suit: Suit
+    let rank: Rank
+    var description: String {
+        return suit.symbol + rank.symbol
+    }
+}
+// Vergleich der Karten
+func ==(lhs: Card, rhs: Card) -> Bool {
+    return lhs.suit == rhs.suit && lhs.rank == rhs.rank
+}
+
 
 
 
-//: ## Testing
 /*
-var rankingCounts = [Ranking : Int]()
-let samples = 100
-for i in 0...samples {
-    let ranking = PokerHand().ranking
-    if rankingCounts[ranking] == nil {
-        rankingCounts[ranking] = 1
-    } else {
-        rankingCounts[ranking]! += 1
+ Enum für Ranking
+ */
+enum Ranking : Int {
+    
+    case Highcard = 0, OnePair = 1, TwoPair = 2, ThreeOfAKind = 3, Straight = 4, Flush = 5, FullHouse = 6, FourOfAKind = 7, StraightFlush = 8, RoyalFlush = 9
+    var description: String {
+        switch self.rawValue {
+        case 0:
+            return "Highcard"
+        case 1:
+            return "One Pair"
+        case 2:
+            return "Two Pair"
+        case 3:
+            return "Three of a kind"
+        case 4:
+            return "Straight"
+        case 5:
+            return "Flush"
+        case 6:
+            return "Full House"
+        case 7:
+            return "Four of a kind"
+        case 8:
+            return "Straight Flush"
+        case 9:
+            return "Royal Flush"
+        default:
+            return "Something went wrong"
+        }
+    }
+}
+// Funktion, die Karten rankt
+func getRankingInt(cards: [Card]) -> [Int] {
```

Comment:
In Swift geht's um typsicheren und ausdrucksstarken Code, das ist von der Philosophie her etwas anders als z.B. Python. Zur Compilezeit sollte immer klar sein womit man es zu tun hat. Wenn du so ein `[Int]` zurückgibst das nur _per Dokumentation_ immer genau 4 Werte enthält die alle für etwas bestimmtest geht das gegen dieses Prinzip. In einem Python Skript ist das nicht schlimm, wenn was schief geht bricht es eben ab, aber eine App darf nicht abstürzen. Versuch also nichts im Code vorauszusetzen was der Compiler nicht weiß.
Hier könntest du z.B. ein `Ranking` zurückgeben und den `Ranking`s jeweils _Associated Values_ geben wie bspw. `case Highcard(Card)`.
Mit `Int`s zu arbeiten die etwas repräsentieren ist insgesamt immer gefährlich, gerade dafür gibt's in Swift Enums, Structs und Protokolle die ebenso mächtig sind wie Klassen um Datenstrukturen zu modellieren.